File: scripts/predict.py
"""Reference predictions with the XGBoost Python package on the GPU.

Reads a CSV with pandas, predicts with an XGBoost JSON model, and writes the
key column, the optional target column, and the prediction, so the result can
be compared with echter_xgb_predict. The model's feature names select the
input columns.
"""

import argparse
import os
import time
from pathlib import Path

import pandas as pd
import xgboost as xgb


def milliseconds(start, end):
    return (end - start) * 1000.0


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("model", help="XGBoost JSON model path")
    parser.add_argument("input", help="CSV input path")
    parser.add_argument(
        "--output",
        help="output CSV path (default: outputs/<input-stem>_xgboost_predictions.csv)",
    )
    parser.add_argument(
        "--key-column",
        help="identifier column written as 'id' (default: the first CSV column)",
    )
    parser.add_argument(
        "--target-column",
        help="optional column written as 'target'",
    )
    return parser.parse_args()


def main():
    args = parse_args()
    output_path = args.output or (
        f"outputs/{Path(args.input).stem}_xgboost_predictions.csv"
    )

    total_start = time.perf_counter()

    # ============================================================
    # CSV READ
    # ============================================================

    read_start = time.perf_counter()
    print("Loading CSV...")
    df = pd.read_csv(args.input)
    read_end = time.perf_counter()

    print(f"Rows: {len(df)}")
    print(f"Columns: {len(df.columns)}")

    # ============================================================
    # MODEL LOAD
    # ============================================================

    model_start = time.perf_counter()
    print("Loading XGBoost model...")
    model = xgb.Booster()
    model.load_model(args.model)
    model.set_param({"device": "cuda"})
    model_end = time.perf_counter()

    features = model.feature_names
    if not features:
        raise RuntimeError("the model has no feature names to select CSV columns")
    print(f"Model expects: {model.num_features()} features")

    missing = [column for column in features if column not in df.columns]
    if missing:
        raise RuntimeError(f"Missing feature columns: {missing}")

    key_column = args.key_column or df.columns[0]
    for column in (key_column, args.target_column):
        if column is not None and column not in df.columns:
            raise RuntimeError(f"Missing column: {column}")

    # ============================================================
    # DMatrix creation
    # ============================================================

    dmatrix_start = time.perf_counter()
    print("Creating DMatrix...")
    dmatrix = xgb.DMatrix(df[features], feature_names=features)
    dmatrix_end = time.perf_counter()

    # ============================================================
    # GPU PREDICTION
    # ============================================================

    predict_start = time.perf_counter()
    print("Running XGBoost prediction on GPU...")
    pred = model.predict(dmatrix)
    predict_end = time.perf_counter()

    # ============================================================
    # CSV WRITE
    # ============================================================

    out = pd.DataFrame({"id": df[key_column]})
    if args.target_column is not None:
        out["target"] = df[args.target_column]
    out["prediction"] = pred

    os.makedirs(os.path.dirname(output_path) or ".", exist_ok=True)
    write_start = time.perf_counter()
    out.to_csv(output_path, index=False)
    write_end = time.perf_counter()

    total_end = time.perf_counter()

    # ============================================================
    # RESULTS
    # ============================================================

    print()
    print("========================================")
    print("XGBoost GPU prediction")
    print("========================================")
    print(f"Model load ms:      {milliseconds(model_start, model_end):.3f}")
    print(f"CSV read ms:        {milliseconds(read_start, read_end):.3f}")
    print(f"DMatrix create ms:  {milliseconds(dmatrix_start, dmatrix_end):.3f}")
    print(f"Prediction ms:      {milliseconds(predict_start, predict_end):.3f}")
    print(f"CSV write ms:       {milliseconds(write_start, write_end):.3f}")
    print(f"Total ms:           {milliseconds(total_start, total_end):.3f}")
    print("----------------------------------------")
    print(f"Rows:               {len(df)}")
    print(f"Input columns:      {len(df.columns)}")
    print(f"Feature columns:    {len(features)}")
    print(f"Output:             {output_path}")
    print("========================================")


if __name__ == "__main__":
    main()

File: apps/predict.cpp
#include <algorithm>
#include <charconv>
#include <chrono>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

import echter.xgb;

namespace
{

long long elapsed_ms(
    std::chrono::steady_clock::time_point start,
    std::chrono::steady_clock::time_point end)
{
    return std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
}

std::size_t parse_index(std::string_view text)
{
    std::size_t value = 0;
    const auto* const end = text.data() + text.size();
    const auto [parsed_end, error] = std::from_chars(text.data(), end, value);
    if (text.empty() || error != std::errc{} || parsed_end != end)
    {
        throw std::invalid_argument("invalid column index '" + std::string(text) + "'");
    }
    return value;
}

std::vector<std::size_t> parse_columns(std::string_view value)
{
    std::vector<std::size_t> columns;
    while (true)
    {
        const std::size_t comma = value.find(',');
        columns.push_back(parse_index(value.substr(0, comma)));
        if (comma == std::string_view::npos)
        {
            return columns;
        }
        value.remove_prefix(comma + 1);
    }
}

void sort_unique(std::vector<std::size_t>& values)
{
    std::sort(values.begin(), values.end());
    values.erase(std::unique(values.begin(), values.end()), values.end());
}

}

int main(int argc, char** argv)
{
    if (argc < 3)
    {
        std::cerr << "usage: echter_xgb_predict <model.json> <input.csv> "
                     "[key-column] [--key-column INDEX] "
                     "[--target-columns INDEX[,INDEX...]] "
                     "[--ignore-columns INDEX[,INDEX...]]\n";
        return 2;
    }

    try
    {
        const std::string model_path = argv[1];
        const std::string input_path = argv[2];
        std::size_t key_column = 0;
        std::vector<std::size_t> target_columns;
        std::vector<std::size_t> excluded_columns;
        int argument = 3;
        if (argument < argc && argv[argument][0] != '-')
        {
            key_column = parse_index(argv[argument++]);
        }
        while (argument < argc)
        {
            const std::string option = argv[argument++];
            if (argument >= argc)
            {
                throw std::invalid_argument("missing value for option " + option);
            }
            const auto columns = parse_columns(argv[argument++]);
            if (option == "--key-column")
            {
                if (columns.size() != 1)
                {
                    throw std::invalid_argument("--key-column expects one index");
                }
                key_column = columns.front();
            }
            else if (option == "--target-columns")
            {
                target_columns.insert(target_columns.end(), columns.begin(), columns.end());
                excluded_columns.insert(excluded_columns.end(), columns.begin(), columns.end());
            }
            else if (option == "--ignore-columns")
            {
                excluded_columns.insert(excluded_columns.end(), columns.begin(), columns.end());
            }
            else
            {
                throw std::invalid_argument("unknown option: " + option);
            }
        }
        const auto total_start = std::chrono::steady_clock::now();

        echter::xgb::XGBoost api;
        auto& regression = api.regression();

        const auto model_start = std::chrono::steady_clock::now();
        regression.load_model(model_path);
        const auto model_end = std::chrono::steady_clock::now();

        const auto read_start = std::chrono::steady_clock::now();
        const auto input = echter::xgb::io::read_csv(input_path);
        const auto read_end = std::chrono::steady_clock::now();

        excluded_columns.push_back(key_column);
        sort_unique(target_columns);
        sort_unique(excluded_columns);
        if (excluded_columns.back() >= input.features())
        {
            throw std::runtime_error(
                "column index " + std::to_string(excluded_columns.back())
                + " is out of range: the CSV has " + std::to_string(input.features())
                + " columns");
        }
        if (input.features() - excluded_columns.size() != regression.num_features())
        {
            throw std::runtime_error(
                "CSV feature count does not match model: model expects "
                + std::to_string(regression.num_features())
                + ", input has "
                + std::to_string(input.features() - excluded_columns.size()));
        }

        const auto input_view = input.view();
        const auto features =
            echter::xgb::io::select_columns(input, excluded_columns);
        const auto predict_start = std::chrono::steady_clock::now();
        const auto predictions = regression.predict(features);
        const auto predict_end = std::chrono::steady_clock::now();

        std::filesystem::create_directories("outputs");
        const std::string output_path =
            "outputs/" + std::filesystem::path(input_path).stem().string()
            + "_predictions.csv";
        const auto write_start = std::chrono::steady_clock::now();
        std::vector<echter::xgb::DeviceColumnarView> output_columns{
            {input_view.data + key_column * input_view.rows, input_view.rows, 1}};
        std::vector<std::string> output_names{"id"};
        for (const auto column : target_columns)
        {
            output_columns.push_back(
                {input_view.data + column * input_view.rows, input_view.rows, 1});
            output_names.push_back("target_" + std::to_string(column));
        }
        output_columns.push_back(predictions.values.view());
        output_names.push_back("prediction");
        echter::xgb::io::write_csv(
            output_path,
            output_columns,
            output_names);
        const auto write_end = std::chrono::steady_clock::now();

        std::cout << "Echter XGBoost prediction\n"
                  << "  model: " << model_path << '\n'
                  << "  input: " << input_path << '\n'
                  << "  output: " << output_path << '\n'
                  << "  rows: " << input.rows() << '\n'
                  << "  input columns: " << input.features() << '\n'
                  << "  feature columns: " << regression.num_features() << '\n'
                  << "  trees: " << regression.num_trees() << '\n'
                  << "  model load ms: " << elapsed_ms(model_start, model_end) << '\n'
                  << "  CSV read ms: " << elapsed_ms(read_start, read_end) << '\n'
                  << "  prediction ms: " << elapsed_ms(predict_start, predict_end) << '\n'
                  << "  CSV write ms: " << elapsed_ms(write_start, write_end) << '\n'
                  << "  total ms: " << elapsed_ms(total_start, write_end) << '\n';
    }
    catch (const std::exception& error)
    {
        std::cerr << "error: " << error.what() << '\n';
        return 1;
    }

    return 0;
}
